### PR TITLE
Backend for bit reversal

### PR DESCRIPTION
**Pull request by maxkurze1**
This PR adds a verilog backend for bit reversal

The following koika:

```
Definition urule  : uaction reg_t empty_ext_fn_t := {{
  write0(r1, >< read0(r1))
}}.
```

can now generate this verilog:

```
// -*- mode: verilog -*-
// Generated by cuttlec from a Kôika module
module reverse(input wire[0:0] CLK, input wire[0:0] RST_N);
	reg[15:0] r1 /* verilator public */ = 16'b0;


	always @(posedge CLK) begin
		if (!RST_N) begin
			r1 <= 16'b0;
		end else begin
			r1 <= {r1[15], r1[14], r1[13], r1[12], r1[11], r1[10], r1[9], r1[8], r1[7], r1[6], r1[5], r1[4], r1[3], r1[2], r1[1], r1[0]};
		end
	end
endmodule
```

### CODE REVIEW

File: ocaml/backends/verilog.ml [hunk 1]
```diff
@@ -227,6 +227,14 @@ let assignment_to_string' (gensym: int ref) (assignment: assignment) =
    | EUnop (fn, arg1) ->
       (match fn with
        | Not _ -> default_left ^ "~" ^ arg1
+       | Rev sz ->
+        let rec loop n = if n = 0
+          then Printf.sprintf "%s[%d]" arg1 n
+          else Printf.sprintf "%s[%d], " arg1 n ^ loop (n - 1) in
+        let rhs = if sz = 0
+            then "{}" (* is that legal for a bits 0 vector? *)
+            else "{" ^ loop (sz-1) ^ "}" in
```
Comment (spacefrogg) [inline ocaml/backends/verilog.ml]: For `n = 0` this should be enough: `x = y`. There is no real vector. So it should not need indexing.

File: tests/reverse.v [hunk 2]
```diff
@@ -0,0 +1,38 @@
+(*! Regression test for signed shifts !*)
```
Comment (spacefrogg) [inline tests/reverse.v]: Copy pasta error. ;)

File: ocaml/backends/verilog.ml (hunk 1 above)
Comment (spacefrogg) [reply ocaml/backends/verilog.ml]: It  should be true for `n = 1` as well.

**Comment by maxkurze1** [comment]
Should be fixed now :)

**Comment by maxkurze1** [comment]
now it looks like this:
```
// -*- mode: verilog -*-
// Generated by cuttlec from a Kôika module
module reverse(input wire[0:0] CLK, input wire[0:0] RST_N);
	reg[0:0] r1 /* verilator public */ = 1'b0;


	always @(posedge CLK) begin
		if (!RST_N) begin
			r1 <= 1'b0;
		end else begin
			r1 <= r1;
		end
	end
endmodule
```

in case of a 1bit vector

File: ocaml/backends/rtl.ml [hunk 3]
```diff
@@ -388,6 +388,14 @@ module Verilog : RTLBackend = struct
     | EUnop (f, e) ->
        (match f with
         | Not _ -> sp "~%a" p e
+        | Rev sz ->
+          let rec loop n = if n = 0
+            then sprintf "%a[%d]" p e n
+            else sprintf "%a[%d], " p e n ^ loop (n - 1) in
```
Comment (spacefrogg) [inline ocaml/backends/rtl.ml:394]: Should we use `sp` here? It looks like it controls operator precedence and adds proper parenthesis.

Comment (maxkurze1) [reply ocaml/backends/rtl.ml:394]: so `sp` is defined as:
```
let sp fmt = ksprintf (fun s -> (lvl, s)) fmt in
```

`ksprintf` does the same as `sprintf` except that it passes it's "output" string as input to its first argument

so its the same as
```
let sp fmt = (lvl, sprintf fmt) in
``` 
However this version won't work as additional arguments to `sp` are not passed on to `sprintf`

Now `sp` generates a tuple, whereas we only want a string to continue the concatenation (which is why I am only using `sprintf`). The tuple is then generated on line 398.
So basically `sp` is only syntax sugar for the smaller cases

Comment (spacefrogg) [reply ocaml/backends/rtl.ml:394]: Ah! Nice.
